students/Lina/Lesson9/donor_models.py

```diff
@@ -0,0 +1,164 @@
+
+    @property
+    def first_name(self):
+        return self.full_name.split()[0]
+
+    @property
+    def last_name(self):
+        return self.full_name.split()[1]
+
+    @property
+    def total_gift_amount(self):
+        return round(sum(self.details["gift_amounts"]), 2)
+
+    @property
+    def total_gift_count(self):
+        return len([amount for amount in self.details["gift_amounts"] if amount > 0])
+
+    @property
+    def avg_gift_amount(self):
+        if self.total_gift_count > 0:
+            avg = round(self.total_gift_amount / self.total_gift_count, 2)
+        else:
+            avg = 0
+        return avg
+
+    def thank_you_letter(self, first_name, gift_amount):
+        #Compose a letter to thank the donor for the donation that they just made.
+        #returns letter in a string
+        letter_layout, sender_template = self.letter_template(first_name, gift_amount)
+
+        letter = "\n".join(("Dear {donor_first_name},",
+                 "{new_paragraph_indent}Thank you for your generous gift of ${donation_amount:.2f} to support the work of ChangeALife. Your",
+                 "donation will remain in your local community and be used by ChangeALife to provide programs",
+                 "and services to those in need.",
+                 "{new_paragraph_indent}On behalf of those whose lives will be impacted, please accept our sincere gratitude. May ",
+                 "you know the deep satisfaction of having made a difference in the lives of others.",
+                 "\n\n",
+                 "Below is a summary of your gift.",
+                 "",
+                 "Amount: ${donation_amount:.2f}",
+                 "Date: {date:%m-%d-%Y}",
+                 sender_template)).format(**letter_layout)
+        return letter
+
+    def letter_template(self, first_name, gift_amount):
+        #Creates a common format for some letters.
+        #Returns letter layout and sender template in a string
+        letter_layout = {"donor_first_name" : first_name,
+                         "donation_amount" : gift_amount,
+                         "sender_name" : "Tiffany Kurnett",
+                         "sender_title" : "President & CEO",
+                         "sender_alignment" : ' '*25,
+                         "new_paragraph_indent" : '\n   ',
+                         "date" : date.today()
+                        }
+        sender_template = "\n".join(("\n\n\n\n{sender_alignment}Sincerely,",
+                          "{sender_alignment}{sender_name}",
+                          "{sender_alignment}{sender_title}",
+                          "\n\n\n"))
+        return letter_layout, sender_template
+
+# DonorCollection class contains information about all donors
+class DonorCollection:
+    #This dictionary database contains the donor names and a history of their donation amounts.
+    def __init__(self):
+        self.donor_db = self.get_donors_history()
```

Comment:
this class should hold reference to your donor objects, so a good data structure could be:
```
self.donor_db = {"<donor_name_str>": <donor_object_instance>}
```